### PR TITLE
case attachments in separate table

### PR DESCRIPTION
**Pull request by snopoke**
@benrudolph

### CODE REVIEW

File: corehq/form_processor/models.py [hunk 1]
```diff
@@ -36,7 +36,21 @@ def is_saved(self):
         return bool(self._get_pk_val())
 
 
-class XFormInstanceSQL(PreSaveHashableMixin, models.Model, AbstractXFormInstance, RedisLockableMixIn, SaveStateMixin):
+class AttachmentMixin(SaveStateMixin):
+    """Requires the model to be linked to the attachments model via the 'attachments' related name.
+    """
+    ATTACHMENTS_RELATED_NAME = 'attachments'
+    def get_attachment(self, attachment_name):
```
Comment (dimagimon) [inline corehq/form_processor/models.py]: E301 expected 1 blank line, found 0

File: corehq/form_processor/models.py [hunk 2]
```diff
@@ -36,7 +36,21 @@ def is_saved(self):
...
+        if hasattr(self, 'unsaved_attachments'):
+            for attachment in self.unsaved_attachments:
+                if attachment.name == attachment_name:
+                    return attachment.read_content()
+        elif self.is_saved():
+            xform_attachment = self.attachments.filter(name=attachment_name).first()
+            return xform_attachment.read_content()
+
+
+class XFormInstanceSQL(PreSaveHashableMixin, models.Model, AbstractXFormInstance, RedisLockableMixIn, AttachmentMixin):
```
Comment (dimagimon) [inline corehq/form_processor/models.py]: E501 line too long (119 > 115 characters)

File: corehq/form_processor/models.py [hunk 3]
```diff
@@ -289,7 +302,7 @@ class XFormPhoneMetadata(jsonobject.JsonObject):
     location = GeoPointProperty()
 
 
-class CommCareCaseSQL(PreSaveHashableMixin, models.Model, AbstractCommCareCase, RedisLockableMixIn, SaveStateMixin):
+class CommCareCaseSQL(PreSaveHashableMixin, models.Model, AbstractCommCareCase, RedisLockableMixIn, AttachmentMixin):
```
Comment (dimagimon) [inline corehq/form_processor/models.py]: E501 line too long (117 > 115 characters)

File: corehq/form_processor/models.py [hunk 4]
```diff
@@ -36,7 +36,22 @@ def is_saved(self):
...
+        if hasattr(self, 'unsaved_attachments'):
+            for attachment in self.unsaved_attachments:
+                if attachment.name == attachment_name:
+                    return attachment.read_content()
+        elif self.is_saved():
+            xform_attachment = self.attachments.filter(name=attachment_name).first()
+            return xform_attachment.read_content()
+
+
+class XFormInstanceSQL(PreSaveHashableMixin, models.Model, RedisLockableMixIn, AttachmentMixin, AbstractXFormInstance):
```
Comment (dimagimon) [inline corehq/form_processor/models.py]: E501 line too long (119 > 115 characters)

File: corehq/form_processor/models.py [hunk 5]
```diff
@@ -289,7 +303,7 @@ class XFormPhoneMetadata(jsonobject.JsonObject):
     location = GeoPointProperty()
 
 
-class CommCareCaseSQL(PreSaveHashableMixin, models.Model, AbstractCommCareCase, RedisLockableMixIn, SaveStateMixin):
+class CommCareCaseSQL(PreSaveHashableMixin, models.Model, RedisLockableMixIn, AttachmentMixin, AbstractCommCareCase):
```
Comment (dimagimon) [inline corehq/form_processor/models.py]: E501 line too long (117 > 115 characters)

**Comment by snopoke** [comment]
:white_check_mark:

**Comment by benrudolph** [comment]
a few comments, nothing blocking. merging :truck:

**Pull request merged**